src/transform/src/demand.rs

```diff
@@ -54,7 +54,14 @@ impl Demand {
         mut columns: HashSet<usize>,
         gets: &mut HashMap<Id, HashSet<usize>>,
     ) {
-        let relation_type = relation.typ();
+        // A valid relation type is only needed for Maps, but we can't borrow
+        // the relation in the corresponding branch of the match statement, since
+        // it is already borrowed mutably.
+        let relation_type = if matches!(relation, MirRelationExpr::Map { .. }) {
+            relation.typ()
+        } else {
+            RelationType::empty()
+        };
```

Comment:
Idiomatically, it feels a bit better to have a `Some(relation.typ())` and a `None` here, to avoid holding invalid data.